File: Java/BSTtoLL.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    void Inorder(TreeNode root, ArrayList<TreeNode> ans)
    {
        if(root == null)
            return;
            
        ans.add(root);
        Inorder(root.left, ans);
        Inorder(root.right, ans);
    }

    public void flatten(TreeNode root) {
        ArrayList<TreeNode> ans = new ArrayList<TreeNode>();
        Inorder(root, ans);
        ans.add(null);

        for(int i = 0; i<ans.size()-1; i++)
        {
            ans.get(i).right = ans.get(i+1);
            ans.get(i).left = null;
        }
    }
}


//Approach-2
class Solution {
  private TreeNode head;
  public void flatten(TreeNode root) {
    if (root == null) return;
    // postorder
    flatten(root.right);
    flatten(root.left);
    root.right = head;
    root.left = null;
    head = root;
  }
}

File: DSA/Tree/BSTtoLL.cpp
void Inorder(Node* root, vector<Node*> &ans)
{
    if(root == NULL)
    {
        return;
    }

    Inorder(root->left, ans);
    ans.push_back(root);
    Inorder(root->right, ans);
}

void convert(vector<Node*> ans)
{
    for(int i = 0; i<ans.size()-1; i++)
    {
        ans[i]->left = NULL;
        ans[i]->right = ans[i+1];
    }

    // Last Node
    ans[ans.size()-1]->left = NULL;
    ans[ans.size()-1]->right = NULL;
}

int main()
{
    vector<Node*> ans;
    Inorder(root, ans);
    convert(ans);
    return ans[0];
}